relayer/chains/cosmos/event_parser.go

```diff
@@ -0,0 +1,240 @@
+			}
+		}
+		if messageInfo == nil {
+			// Not an IBC message, don't need to log here
+			continue
+		}
+		if messageType == "" {
+			ccp.log.Error("unexpected ibc message parser state: message info is populated but type is empty",
+				zap.String("chain_id", ccp.ChainProvider.ChainId()),
+				zap.Any("message", messageInfo),
+			)
+			continue
+		}
+		messages = append(messages, ibcMessage{
+			messageType: messageType,
+			messageInfo: messageInfo,
+		})
+	}
+
+	return messages
+}
+
+func (ccp *CosmosChainProcessor) parseClientInfo(attributes []sdk.Attribute) clientInfo {
+	res := clientInfo{}
+	for _, attr := range attributes {
+		switch attr.Key {
+		case "client_id":
+			res.clientID = attr.Value
+		case "consensus_height":
+			revisionSplit := strings.Split(attr.Value, "-")
+			if len(revisionSplit) != 2 {
+				ccp.log.Error("error parsing client consensus height",
+					zap.String("chain_id", ccp.ChainProvider.ChainId()),
+					zap.String("client_id", res.clientID),
+					zap.String("value", attr.Value),
+				)
+				continue
+			}
+			revisionNumberString := revisionSplit[0]
+			revisionNumber, err := strconv.ParseUint(revisionNumberString, 10, 64)
+			if err != nil {
+				ccp.log.Error("error parsing client consensus height revision number",
+					zap.String("chain_id", ccp.ChainProvider.ChainId()),
+					zap.Error(err),
+				)
+				continue
+			}
+			revisionHeightString := revisionSplit[1]
+			revisionHeight, err := strconv.ParseUint(revisionHeightString, 10, 64)
+			if err != nil {
+				ccp.log.Error("error parsing client consensus height revision height",
+					zap.String("chain_id", ccp.ChainProvider.ChainId()),
+					zap.Error(err),
+				)
+				continue
+			}
+			res.consensusHeight = clienttypes.Height{
+				RevisionNumber: revisionNumber,
+				RevisionHeight: revisionHeight,
+			}
+		}
+	}
+	return res
+}
+
+func (ccp *CosmosChainProcessor) parsePacketInfo(attributes []sdk.Attribute) packetInfo {
+	res := packetInfo{packet: chantypes.Packet{}}
+	for _, attr := range attributes {
+		var err error
+		switch attr.Key {
+		case "packet_sequence":
+			res.packet.Sequence, err = strconv.ParseUint(attr.Value, 10, 64)
+			if err != nil {
+				ccp.log.Error("error parsing packet sequence",
+					zap.String("chain_id", ccp.ChainProvider.ChainId()),
+					zap.String("value", attr.Value),
+					zap.Error(err),
+				)
+				continue
+			}
+		case "packet_timeout_timestamp":
+			res.packet.TimeoutTimestamp, err = strconv.ParseUint(attr.Value, 10, 64)
+			if err != nil {
+				ccp.log.Error("error parsing packet timestamp",
+					zap.String("chain_id", ccp.ChainProvider.ChainId()),
+					zap.Uint64("sequence", res.packet.Sequence),
+					zap.String("value", attr.Value),
+					zap.Error(err),
+				)
+				continue
+			}
+		case "packet_data":
+			res.packet.Data = []byte(attr.Value)
+		case "packet_data_hex":
+			data, err := hex.DecodeString(attr.Value)
+			if err == nil {
+				ccp.log.Error("error parsing packet data",
```

Comment:
Should log and continue on `err != nil`.

Reply:
I feel we use `continue` and for loops too much here and elsewhere. It lends itself to this exact problem, forgetting to continue. 

I think there's a pattern where we can abstract these parts into functions/methods with `return` values. That way, the compiler forces us to remember to return something. The compiler does not enforce continue in this manner.

E.g. 

```go
func parseValue(attr sdk.Attribute, packet *packetInfo) error {
  switch attr.Key {
  case "packet_sequence":
    packet.Sequence, err := strconv.ParseUint(attr.Value, 10, 64)
    if err != nil {
      return fmt.Errorf("parse sequence: %w", err)
    // ...
}

for _, attr := range attributes {
   parseValue(attr.Key, &info)
   // ...
```

Another option:

It may be less performant, but ibctest use a helper function which abstracts the for loop for this messy, loose event/attribute data structure. The calling code is much cleaner imo. https://github.com/strangelove-ventures/ibctest/blob/60cd4d1fb280e3e1d39e0cc77612e200073ef8e3/chain/cosmos/cosmos_chain.go#L152

Reply:
> Should log and continue on err != nil.

good catch.

>I think there's a pattern where we can abstract these parts into functions/methods with return values.

done. I think I prefer the return pattern over unnecessary iterations here for performance reasons, but I do like how nice the API is for ibctest.